File: src/robot_racing_car/nodes/autonomous_controller.py
#!/usr/bin/env python

import rospy
import cv2
import numpy as np
import tensorflow as tf
import h5py
import constants

from std_msgs.msg import UInt8
from sensor_msgs.msg import CompressedImage
from keras.models import load_model
from random import uniform


class AutonomousController:

    model = load_model("../model/model.h5")
    graph = tf.get_default_graph()
    last_image = None
    is_steering = False

    def camera_callback(self, message):
        np_arr = np.fromstring(message.data, np.uint8)
        image = cv2.imdecode(np_arr, cv2.IMREAD_COLOR)
        small_image = np.empty((image.shape[1] / constants.IMAGE_SCALE, image.shape[0] / constants.IMAGE_SCALE))
        image = cv2.resize(image, small_image.shape, interpolation=cv2.INTER_LINEAR)

        self.last_image = np.asarray(image)

        with self.graph.as_default():
            self.last_image = self.last_image.reshape((1,) + self.last_image.shape)
            prediction = self.model.predict(self.last_image)

            steering_command = np.argmax(prediction[0][0])
            throttle_value = prediction[1][0][0]

            rospy.logdebug("Predicted steering command: " + str(prediction[0][0]))
            rospy.logdebug("Predicted throttle: " + str(throttle_value))

            self.robot_car_steering_speed_publisher.publish(int(throttle_value))

            if steering_command != constants.MOVE_FORWARD_COMMAND:
                is_steering = True
                self.robot_car_steering_publisher.publish(steering_command)
                is_steering = False

    def steering_callback(self, callback_argument):
        if not self.is_steering:
            self.robot_car_steering_publisher.publish(constants.MOVE_FORWARD_COMMAND)

    def __init__(self):
        rospy.init_node("autonomous_controller", log_level=rospy.DEBUG)

        rospy.Subscriber("/raspicam_node/image/compressed", CompressedImage, self.camera_callback, queue_size=1)

        self.robot_car_steering_publisher = rospy.Publisher("robot_car_steering", UInt8, queue_size=1)
        self.robot_car_steering_speed_publisher = rospy.Publisher("robot_car_speed", UInt8, queue_size=1)

        rospy.Timer(rospy.Duration(0.1), self.steering_callback)


if __name__ == "__main__":
    controller = AutonomousController()

    rospy.spin()

File: src/autonomous_controller/src/AutonomousController.cpp
#include <ros/ros.h>
#include <sensor_msgs/Joy.h>
#include <sensor_msgs/CompressedImage.h>
#include <sensor_msgs/Image.h>
#include "std_msgs/String.h"

#include <cv_bridge/cv_bridge.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

using namespace std_msgs;
using namespace sensor_msgs;

ros::Publisher actuatorPublisher;

void cameraCallback(const CompressedImageConstPtr& image)
{
	ROS_INFO("Callback!");
	
	cv::Mat cv_image = cv::imdecode(cv::Mat(image->data), CV_LOAD_IMAGE_COLOR);
	cv::Mat resized_cv_image;	
	cv::resize(cv_image, resized_cv_image, cv::Size(), 0.25, 0.25);
}

int main(int argc, char* argv[])
{ 
	ros::init(argc, argv, "autonomous_controller");

	ros::NodeHandle nodeHandle;
	
	ros::Subscriber cameraSubscriber = nodeHandle.subscribe<CompressedImage>("/raspicam_node/image/compressed", 1, cameraCallback);
	
	actuatorPublisher = nodeHandle.advertise<String>("/autonomous", 1);
	
	ROS_INFO("Autonomous controller node started.");

	ros::spin();
	
	return 0;
}
